src/lib/tvgSceneImpl.h

```diff
@@ -81,12 +86,21 @@ struct Scene::Impl
 
     bool needComposition(uint32_t opacity)
     {
+        // masking may require composition
+        auto compMethod = scene->composite(nullptr);
+        if (paints.count > 0 && compMethod != CompositeMethod::None && compMethod != CompositeMethod::ClipPath)
+            return true;
+
         //Half translucent requires intermediate composition.
-        if (opacity == 255 || opacity == 0) return false;
+        if (opacity == 255 || opacity == 0)
+            return false;
 
         //If scene has several children or only scene, it may require composition.
-        if (paints.count > 1) return true;
-        if (paints.count == 1 && (*paints.data)->identifier() == TVG_CLASS_ID_SCENE) return true;
+        if (paints.count > 1)
```

Comment:
Shouldn't it be modified with `if (paints.count > 0)`? (no method check)
Are there any exception cases?

Reply:
I'm not sure what do you mean. I didn;t modify this part, just the new lines are added (sorry for that).
Sure we can check whether there are some paints on the very beginning of this function, but I guess it's not what you meant...

Reply:
Sorry for the confusion.
What I meant was to perform composition unconditionally if there is paint in the scene.
(However, this method doesn't seem appropriate because it will return true even in unnecessary composition cases.)

The needComposition() function is getting a bit complicated. How about clean up priorities a bit?
I think at least opacity==0 should be on top in all cases.
like this?
```cpp
            if (opacity == 0) return false;

        auto compMethod = scene->composite(nullptr);
        if (paints.count > 0 && compMethod != CompositeMethod::None && compMethod != CompositeMethod::ClipPath) return true;

        //Half translucent requires intermediate composition.
        if (opacity == 255) return false;

        //If scene has several children or only scene, it may require composition.
        if (paints.count > 1) return true;
        if (paints.count == 1 && (*paints.data)->identifier() == TVG_CLASS_ID_SCENE) return true;
        return false;
    }
```

+)
nit : please remove the newline after the if statement.